File: frontend/flutter_app/lib/models/receta_ingrediente.dart
import 'ingrediente.dart';
import 'unidad_medida.dart';

/// Relación entre una receta y uno de sus ingredientes,
/// con la cantidad necesaria para elaborarla.
///
/// [unidadId] y [unidad] son la unidad de medida (catálogo global) en la que
/// se expresa [cantidadNecesaria] (Fase 2 "unidades-medida", PR3). Pueden
/// venir `null` si el backend todavía no los expone (compatibilidad con
/// despliegues anteriores a esta fase); en ese caso se asume la propia
/// unidad del ingrediente.
class RecetaIngrediente {
  final int        id;
  final int        recetaId;
  final Ingrediente ingrediente;
  final double     cantidadNecesaria;
  final int?       unidadId;
  final UnidadMedida? unidad;

  const RecetaIngrediente({
    required this.id,
    required this.recetaId,
    required this.ingrediente,
    required this.cantidadNecesaria,
    this.unidadId,
    this.unidad,
  });

  factory RecetaIngrediente.fromJson(Map<String, dynamic> json) =>
      RecetaIngrediente(
        id:                json['id']                as int,
        recetaId:          json['recetaId']          as int,
        ingrediente:       Ingrediente.fromJson(json['ingrediente'] as Map<String, dynamic>),
        cantidadNecesaria: (json['cantidadNecesaria'] as num).toDouble(),
        unidadId:          json['unidadId']          as int?,
        unidad:            json['unidad'] != null
            ? UnidadMedida.fromJson(json['unidad'] as Map<String, dynamic>)
            : null,
      );

  Map<String, dynamic> toJson() => {
        'id':                id,
        'recetaId':          recetaId,
        'ingrediente':       ingrediente.toJson(),
        'cantidadNecesaria': cantidadNecesaria,
        if (unidadId != null) 'unidadId': unidadId,
        if (unidad != null) 'unidad': unidad!.toJson(),
      };

  RecetaIngrediente copyWith({
    int?         id,
    int?         recetaId,
    Ingrediente? ingrediente,
    double?      cantidadNecesaria,
    int?         unidadId,
    UnidadMedida? unidad,
  }) =>
      RecetaIngrediente(
        id:                id                ?? this.id,
        recetaId:          recetaId          ?? this.recetaId,
        ingrediente:       ingrediente       ?? this.ingrediente,
        cantidadNecesaria: cantidadNecesaria ?? this.cantidadNecesaria,
        unidadId:          unidadId          ?? this.unidadId,
        unidad:            unidad            ?? this.unidad,
      );
}

File: backend/src/main/java/com/ares/backend/entity/RecetaIngrediente.java
package com.ares.backend.entity;

import jakarta.persistence.*;
import lombok.*;

/**
 * Entidad intermedia que relaciona una receta con sus ingredientes necesarios.
 * Especifica la cantidad exacta de cada ingrediente requerida.
 *
 * @author Ares
 * @version 1.0
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@ToString
@Table(name = "receta_ingredientes")
public class RecetaIngrediente {

    /**
     * Identificador único de la relación.
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Receta que requiere el ingrediente.
     */
    @ManyToOne
    @JoinColumn(name = "receta_id", nullable = false)
    private Receta receta;

    /**
     * Ingrediente necesario para la receta.
     */
    @ManyToOne
    @JoinColumn(name = "ingrediente_id", nullable = false)
    private Ingrediente ingrediente;

    /**
     * Cantidad necesaria del ingrediente para elaborar la receta, expresada
     * en {@link #unidad} (no necesariamente la unidad del ingrediente — ver
     * {@link com.ares.backend.service.RecetaService} para la conversión).
     */
    @Column(nullable = false)
    private Double cantidadNecesaria;

    /**
     * Unidad de medida en la que se expresa {@link #cantidadNecesaria}
     * (catálogo global {@link UnidadMedida}). Por defecto es la propia
     * unidad base del ingrediente, pero puede diferir siempre que comparta
     * {@code tipo} — {@code RecetaService} valida esa compatibilidad al
     * crear/actualizar la receta (Fase 2 "unidades-medida", PR3, decisión
     * D4). No se marca {@code nullable = false} a nivel de anotación JPA por
     * el mismo motivo que {@link Ingrediente#getUnidadBase()}: el esquema de
     * test (H2, {@code ddl-auto=create-drop}) no ejecuta la migración V5,
     * que sí impone NOT NULL en BD real.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "unidad_id")
    private UnidadMedida unidad;

    /**
     * Constructor con parámetros para crear una relación receta-ingrediente.
     *
     * @param receta Receta que requiere el ingrediente
     * @param ingrediente Ingrediente necesario
     * @param cantidadNecesaria Cantidad requerida
     */
    public RecetaIngrediente(Receta receta, Ingrediente ingrediente, Double cantidadNecesaria) {
        this.receta = receta;
        this.ingrediente = ingrediente;
        this.cantidadNecesaria = cantidadNecesaria;
    }
}